File: server/model/sftp.go
package model

import "time"

// SftpEntry represents a file or directory entry. Field names use snake_case
// to match the frontend types/sftp.ts contract.
type SftpEntry struct {
	Name    string `json:"name"`               // base name without path
	Path    string `json:"path"`               // absolute POSIX path (no trailing slash)
	IsDir   bool   `json:"is_dir"`             // true for directories
	Size    int64  `json:"size"`               // bytes; 0 for directories
	ModTime string `json:"mod_time"`           // RFC 3339 timestamp
	Mode    string `json:"mode,omitempty"`     // Unix permission string e.g. "rwxr-xr-x" (optional)
}

// SftpTreeNode is a tree entry used by the recursive tree endpoint.
type SftpTreeNode struct {
	SftpEntry
	Children []SftpTreeNode `json:"children,omitempty"`
}

// SftpServer describes a connection target.
type SftpServer struct {
	ID       string `json:"id"`       // "local" or a profile ID
	Name     string `json:"name"`     // display name
	Host     string `json:"host"`     // hostname
	Port     int    `json:"port"`     // port; 0 for local
	Username string `json:"username"` // login user
}

// TransferTask represents an asynchronous file transfer.
type TransferTask struct {
	ID           string `json:"id"`                      // task ID
	FileName     string `json:"file_name"`               // file name
	Direction    string `json:"direction"`               // "upload" | "download"
	Size         int64  `json:"size"`                    // total bytes
	Transferred  int64  `json:"transferred"`             // bytes transferred
	Status       string `json:"status"`                  // queued | transferring | completed | failed | cancelled
	Speed        int64  `json:"speed"`                   // bytes/sec
	StartedAt    int64  `json:"started_at"`              // Unix milliseconds
	FinishedAt   *int64 `json:"finished_at,omitempty"`   // completion timestamp
	ErrorMessage string `json:"error_message,omitempty"` // failure reason
}

// --- Request / Response DTOs ---

type SftpCreateSessionRequest struct {
	ProfileID string `json:"profile_id"`
}

type SftpCreateSessionResponse struct {
	SessionID string `json:"session_id"`
	Status    string `json:"status"`
}

type SftpSessionInfo struct {
	ID        string    `json:"id"`
	ProfileID string    `json:"profile_id"`
	Status    string    `json:"status"`
	Error     string    `json:"error,omitempty"`
	CreatedAt time.Time `json:"created_at"`
}

type SftpListResponse struct {
	Path    string      `json:"path"`
	Entries []SftpEntry `json:"entries"`
}

type SftpTreeResponse struct {
	Path    string          `json:"path"`
	Entries []SftpTreeNode  `json:"entries"`
}

type SftpMkdirRequest struct {
	Path string `json:"path"`
}

type SftpRenameRequest struct {
	OldPath string `json:"old_path"`
	NewPath string `json:"new_path"`
}

type SftpDeleteRequest struct {
	Paths []string `json:"paths"`
}

type SftpDeleteResponse struct {
	Deleted int `json:"deleted"`
	Failed  int `json:"failed"`
}

type SftpDownloadRequest struct {
	Paths []string `json:"paths"`
}

type SftpDownloadResponse struct {
	Tasks       []TransferTask `json:"tasks"`
	DownloadURL string         `json:"download_url"`
}

type SftpUploadResponse struct {
	Tasks []TransferTask `json:"tasks"`
}

type SftpCancelTransferResponse struct {
	ID     string `json:"id"`
	Status string `json:"status"`
}

File: web/src/types/sftp.ts
/** SFTP file entry — mirrors backend model/sftp.go SftpEntry.
 *  Field names use snake_case to match the backend contract. */
export interface SftpEntry {
  name: string
  /** Absolute POSIX path (without trailing slash). */
  path: string
  is_dir: boolean
  size: number // bytes; 0 for directories
  mod_time: string // RFC 3339 timestamp
  /** Unix permission hint, e.g. "rwxr-xr-x". Optional. */
  mode?: string
}

/** Tree node for recursive tree endpoint responses. */
export interface SftpTreeNode extends SftpEntry {
  children?: SftpTreeNode[]
}

export type TransferDirection = 'upload' | 'download'
export type TransferStatus = 'queued' | 'transferring' | 'completed' | 'failed' | 'cancelled'

export interface TransferTask {
  id: string
  file_name: string
  direction: TransferDirection
  size: number
  transferred: number
  status: TransferStatus
  /** bytes/sec */
  speed: number
  started_at: number // Unix milliseconds
  finished_at?: number
  error_message?: string
}

/** A connected SFTP target server. */
export interface SftpServer {
  id: string
  name: string
  host: string
  port: number
  username: string
}

/** Response type for POST /api/sftp/sessions */
export interface SftpCreateSessionResponse {
  session_id: string
  status: string
}

/** Response type for GET /api/sftp/sessions/{id}/list */
export interface SftpListResponse {
  path: string
  entries: SftpEntry[]
}

/** Response type for GET /api/sftp/sessions/{id}/tree */
export interface SftpTreeResponse {
  path: string
  entries: SftpTreeNode[]
}

/** Response type for POST /api/sftp/sessions/{id}/upload */
export interface SftpUploadResponse {
  tasks: TransferTask[]
}

/** Response type for POST /api/sftp/sessions/{id}/download */
export interface SftpDownloadResponse {
  tasks: TransferTask[]
  download_url: string
}

/** Response type for POST /api/sftp/sessions/{id}/delete */
export interface SftpDeleteResponse {
  deleted: number
  failed: number
}

/** Conflict-resolution strategy for cross-session transfers. Mirrors the
 *  backend model.ConflictResolution enum. */
export type ConflictResolution = 'ask' | 'overwrite' | 'rename' | 'skip'

/** A single file collision detected before a transfer. Mirrors backend
 *  model.SftpConflictInfo. */
export interface SftpConflictInfo {
  source_path: string
  dest_path: string
  source_size: number
  dest_size: number
}

/** Response type for POST /api/sftp/transfer.
 *  On success: task_id/method/tasks are populated.
 *  On conflict (HTTP 409): conflicts is populated, task_id is empty. */
export interface SftpTransferResponse {
  task_id?: string
  method?: string
  tasks?: TransferTask[]
  conflicts?: SftpConflictInfo[]
}

/* ─── Built-in editor types ─── */

export type LineEnding = 'lf' | 'crlf'

/** Response type for GET /api/sftp/sessions/{id}/file?path=...
 *  Guards: backend rejects files >10MB (413), binary (415), non-UTF-8 (415). */
export interface SftpFileReadResponse {
  path: string
  content: string
  size: number
  /** RFC 3339Nano timestamp; used as the optimistic-lock token on save. */
  mod_time: string
  /** Monaco language id, e.g. "shell", "json", "nginx", "plaintext". */
  language: string
  line_ending: LineEnding
  /** True when the file's owner-write bit is unset. */
  read_only: boolean
}

/** Request body for PUT /api/sftp/sessions/{id}/file?path=... */
export interface SftpFileWriteRequest {
  content: string
  /** Must match the server's current ModTime; mismatch → 409 FILE_MODIFIED. */
  expected_mod_time: string
  line_ending?: LineEnding
}

/** Response type for PUT /api/sftp/sessions/{id}/file. The new mod_time
 *  becomes the optimistic-lock token for the next save. */
export interface SftpFileWriteResponse {
  path: string
  size: number
  mod_time: string
}
